epoch-verifier/src/verifications/donations.ts

```diff
@@ -1,72 +1,111 @@
-import { assertAll, assertEq } from "./utils";
+import { sqrt, assertAll, assertEq, multiplyFloatByBigInt, SCALE_FACTOR } from "./utils";
 
-import { Context, allocationsByUser, individualDonationsByProposals, rewardsByProject } from "../data/context";
+import { Context, allocationsByUser, individualDonationsSummedByProjects, individualDonationsAggregatedByProjectsWithUQs, rewardsByProject } from "../data/context";
 import { Address, Reward } from "../data/models";
 import { VerificationResult } from "../runner";
 
-const PROPOSALS_NO = 30
+const PROPOSALS_NO = 30;
 
-function getThreshold(individualAllocations: Map<Address, bigint>): bigint {
-  const allocationsSum = Array.from(individualAllocations.entries()).reduce((acc, [_, val]) => acc + val, BigInt(0))
-  return allocationsSum / BigInt(PROPOSALS_NO)
+
+function _groupAllocationsByProjects(aggregatedDonations: Map<Address, UserDonation[]>): [Map<Address, bigint>, bigint] {
+  const result = new Map<Address, bigint>();
+  let totalPlainQF = BigInt(0);
+
+  aggregatedDonations.forEach((values, project) => {
+    const sumOfSqrt = values.reduce((sum, value) => sum + sqrt(multiplyFloatByBigInt(value.uq, value.amount)), BigInt(0));
+    const resultValue = sumOfSqrt ** BigInt(2);
+
+    totalPlainQF += resultValue;
+
+    result.set(project, resultValue);
+  });
```

Comment:
Using `.reduce` should allow creating of a pure function instead of `.forEach` w/ modification of outside variable.